src/base/emu-i386/simx86/codegen-sim.c

```diff
@@ -2058,6 +2058,9 @@ static unsigned int Gen_sim(const IGen *IG, dosaddr_t mem_ref)
 	case O_POP1:
 		GTRACE0("O_POP1");
 		SR1.d = CPULONG(Ofs_ESP);
+#ifdef STACK_WRAP_MP	/* mask after incrementing */
+		SR1.d &= CPULONG(Ofs_STACKM);
```

Comment:
Isn't this redundant?
I suppose if you do proper
masking on POP3, then you
can relatively easily assume
it is still masked on input?

Reply:
No, we always need to mask before the first pop in a pop1 pop2 pop2 pop3 sequence, this mask is simply moved up from the beginning of pop2 (compare pop1/pop2/pop3 concatenated to O_POP to see). With STACK_WRAP_MP you get then 3 masks, without it you get 2 (the very last one didn't get a mask). Example: imagine ESP=0xffff1234 here with 16-bit code and stackmask=0xffff, you want to end up with ESP=0xffff1230, but SR1.d = 0x1234, then 0x1232, then 0x1230, and then it gets merged again.

Where STACK_WRAP_MP is needed is e.g. the same scenario but with esp=0xfffe on input. SR1.d becomes
0x10000 without the post-masking and with `SR1.d |= (CPULONG(Ofs_ESP) & ~stackm);`  it's then or'ed with 0! So you end up with ESP=0x10000 instead of ESP=0!

Reply:
Well, yes although in this
case it shouldn't be under
`ifdef STACK_WRAP_MP`.
As I understand it, this ifdef
is only needed to catch
overflow problems, not
for the essential things
like 16bit ESP access.